pynumdiff/kalman_smooth/_kalman_smooth.py

```diff
@@ -6,638 +6,255 @@
 ####################
 # Helper functions #
 ####################
-
-
-def __kalman_forward_update__(xhat_fm, P_fm, y, u, A, B, C, R, Q):
-    """
-    :param xhat_fm:
-    :param P_fm:
-    :param y:
-    :param u:
-    :param A:
-    :param B:
-    :param C:
-    :param R:
-    :param Q:
-    :return:
-    """
-    I = np.array(np.eye(A.shape[0]))
-    gammaW = np.array(np.eye(A.shape[0]))
-
-    K_f = P_fm@C.T@np.linalg.pinv(C@P_fm@C.T + R)
-
-    if y is not None and not np.isnan(y):
-        xhat_fp = xhat_fm + K_f@(y - C@xhat_fm)
-        P_fp = (I - K_f@C)@P_fm
-        xhat_fm = A@xhat_fp + B@u
-        P_fm = A@P_fp@A.T + gammaW@Q@gammaW.T
```

Comment:
The order of these calls is inverted from how I tend to think about the KF. You can start on either the prediction or the update side, but I start on prediction in the paper, and so do the standard Wikipedia and many other pseudocodes for the KF. Thus, your `_fm` variables here contain the *next* time step's prediction, but they're indexed at this time step, which causes confusion when you go do your RTS implementation.